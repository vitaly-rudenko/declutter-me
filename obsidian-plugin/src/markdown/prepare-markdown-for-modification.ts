import { isEmpty, isLogicalSeparator } from "./markdown-tools.js"

export function prepareMarkdownForModification(input: {
  markdown: string,
  section?: string
}): string {
  const { markdown, section } = input

  const lines = markdown.split('\n')

  if (section) {
    if (!lines.includes(section)) {
      if (lines.every(line => isEmpty(line))) {
        lines[0] = section
        lines[1] = ''
        lines[2] = ''
        lines[3] = ''
      } else {
        if (!isEmpty(lines[lines.length - 1])) {
          lines.push('')
        }

        lines.push(section, '', '', '')
      }
    } else {
      const sectionIndex = lines.indexOf(section)
      const nextLogicalSeparatorIndex = lines.findIndex((line, index) => index > sectionIndex && isLogicalSeparator(line))
      const sectionHasContent = nextLogicalSeparatorIndex === -1
        ? lines.some((line, index) => index > sectionIndex && !isEmpty(line))
        : lines.some((line, index) => index > sectionIndex && index < nextLogicalSeparatorIndex && !isEmpty(line))

      if (!sectionHasContent) {
        while (lines[sectionIndex + 3] === undefined || !isEmpty(lines[sectionIndex + 3])) {
          lines.splice(sectionIndex + 1, 0, '')
        }
      }
    }
  } else {
    const lastNonEmptyLineIndex = lines.findLastIndex((line) => !isEmpty(line))
    if (lastNonEmptyLineIndex !== -1 && isLogicalSeparator(lines[lastNonEmptyLineIndex])) {
      while (lines[lastNonEmptyLineIndex + 3] === undefined || !isEmpty(lines[lastNonEmptyLineIndex + 3])) {
        lines.splice(lastNonEmptyLineIndex + 1, 0, '')
      }
    }
  }

  return lines.join('\n')
}