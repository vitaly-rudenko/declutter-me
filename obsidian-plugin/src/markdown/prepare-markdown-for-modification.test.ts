import { prepareMarkdownForModification } from "./prepare-markdown-for-modification"

describe('prepareMarkdownForModification()', () => {
  it('prepares empty markdown for modification', () => {
    expect(prepareMarkdownForModification({ markdown: '' })).toEqual('')
  })

  it('appends section when missing', () => {
    expect(prepareMarkdownForModification({
      markdown: '',
      section: '# section'
    })).toEqual('# section\n\n\n')

    expect(prepareMarkdownForModification({
      markdown: '# section before\n',
      section: '# section'
    })).toEqual('# section before\n\n# section\n\n\n')
  })

  it('adds missing newlines to the section', () => {
    expect(prepareMarkdownForModification({
      markdown: '# section before\n# section\n# section after',
      section: '# section'
    })).toEqual('# section before\n# section\n\n\n\n# section after')
  })

  it('adds missing newlines after properties', () => {
    expect(prepareMarkdownForModification({
      markdown: '---\nhello: world\n---',
    })).toEqual('---\nhello: world\n---\n\n\n')
  })

  it('correctly appends section after properties', () => {
    expect(prepareMarkdownForModification({
      markdown: '---\nhello: world\n---',
      section: '# section'
    })).toEqual('---\nhello: world\n---\n\n# section\n\n\n')
  })

  it('keeps markdown unmodified when no changes are needed', () => {
    expect(prepareMarkdownForModification({
      markdown: 'hello world',
    })).toEqual('hello world')

    // Already has data in the section
    expect(prepareMarkdownForModification({
      markdown: '# section\nhello world',
      section: '# section'
    })).toEqual('# section\nhello world')

    // Already has data after properties
    expect(prepareMarkdownForModification({
      markdown: '---\nhello: world\n---\nhello world',
    })).toEqual('---\nhello: world\n---\nhello world')
  })
})