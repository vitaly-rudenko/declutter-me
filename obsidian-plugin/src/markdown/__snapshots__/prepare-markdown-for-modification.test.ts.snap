// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareMarkdownForModification() [permutations]:  + section "# section" 1`] = `
"# section


"
`;

exports[`prepareMarkdownForModification() [permutations]:  w/o section 1`] = `""`;

exports[`prepareMarkdownForModification() [permutations]: # section before<br># section<br># section after + section "# section" 1`] = `
"# section before
# section



# section after"
`;

exports[`prepareMarkdownForModification() [permutations]: # section before<br># section<br># section after w/o section 1`] = `
"# section before
# section
# section after


"
`;

exports[`prepareMarkdownForModification() [permutations]: # section before<br># section<br><br># section after + section "# section" 1`] = `
"# section before
# section



# section after"
`;

exports[`prepareMarkdownForModification() [permutations]: # section before<br># section<br><br># section after w/o section 1`] = `
"# section before
# section

# section after


"
`;

exports[`prepareMarkdownForModification() [permutations]: # section before<br># section<br><br><br># section after + section "# section" 1`] = `
"# section before
# section



# section after"
`;

exports[`prepareMarkdownForModification() [permutations]: # section before<br># section<br><br><br># section after w/o section 1`] = `
"# section before
# section


# section after


"
`;

exports[`prepareMarkdownForModification() [permutations]: # section<br>hello world + section "# section" 1`] = `
"# section
hello world"
`;

exports[`prepareMarkdownForModification() [permutations]: # section<br>hello world w/o section 1`] = `
"# section
hello world"
`;

exports[`prepareMarkdownForModification() [permutations]: # section<br>hello world<br> + section "# section" 1`] = `
"# section
hello world
"
`;

exports[`prepareMarkdownForModification() [permutations]: # section<br>hello world<br> w/o section 1`] = `
"# section
hello world
"
`;

exports[`prepareMarkdownForModification() [permutations]: <br> + section "# section" 1`] = `
"# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br> w/o section 1`] = `
"
"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br> + section "# section" 1`] = `
"# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br> w/o section 1`] = `
"


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br># section before<br><br><br># section<br><br><br># section after<br><br><br> + section "# section" 1`] = `
"


# section before


# section



# section after


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br># section before<br><br><br># section<br><br><br># section after<br><br><br> w/o section 1`] = `
"


# section before


# section


# section after


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br># section<br><br><br>hello world<br><br><br> + section "# section" 1`] = `
"


# section


hello world


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br># section<br><br><br>hello world<br><br><br> w/o section 1`] = `
"


# section


hello world


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br>---<br>hello: world<br>---<br><br><br> + section "# section" 1`] = `
"


---
hello: world
---



# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br>---<br>hello: world<br>---<br><br><br> w/o section 1`] = `
"


---
hello: world
---


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br>hello world<br><br><br> + section "# section" 1`] = `
"


hello world



# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: <br><br><br>hello world<br><br><br> w/o section 1`] = `
"


hello world


"
`;

exports[`prepareMarkdownForModification() [permutations]: ---<br>hello: world<br>--- + section "# section" 1`] = `
"---
hello: world
---

# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: ---<br>hello: world<br>--- w/o section 1`] = `
"---
hello: world
---


"
`;

exports[`prepareMarkdownForModification() [permutations]: ---<br>hello: world<br>---<br> + section "# section" 1`] = `
"---
hello: world
---

# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: ---<br>hello: world<br>---<br> w/o section 1`] = `
"---
hello: world
---


"
`;

exports[`prepareMarkdownForModification() [permutations]: hello world + section "# section" 1`] = `
"hello world

# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: hello world w/o section 1`] = `"hello world"`;

exports[`prepareMarkdownForModification() [permutations]: hello world<br> + section "# section" 1`] = `
"hello world

# section


"
`;

exports[`prepareMarkdownForModification() [permutations]: hello world<br> w/o section 1`] = `
"hello world
"
`;
